crates/noirc_frontend/src/monomorphisation/mod.rs

```diff
@@ -465,13 +465,20 @@ impl Monomorphiser {
         &self,
         meta: FuncMeta,
         arguments: Vec<ast::Expression>,
+        arg_ids: Vec<node_interner::ExprId>,
     ) -> ast::Expression {
         let attribute = meta.attributes.expect("all builtin functions must contain an attribute which contains the function name which it links to");
         let opcode = attribute
             .builtin()
             .expect("ice: function marked as a builtin, but attribute kind does not match this");
 
-        ast::Expression::CallBuiltin(ast::CallBuiltin { opcode, arguments })
+        if opcode == "arraylen" {
```

Comment:
I believe this function is called when the frontend is parsing the code, which does not work for all cases.
For instance in the array_len examples, it does not work with this call_me method:
```
fn call_me(b: [Field]) -> Field{
    len_plus_1(b)
}
fn main(len3: [u8; 3], len4: [Field; 4]) {
     constrain len_plus_1(len3) == 4;
     constrain len_plus_1(len4) == 5;
     constrain add_lens(len3, len4) == 7;
     constrain call_me(len4) == 5;
}
```
If you look at the implementation of std:len() I did in the SSA pass, I process the functions in the order they are called and keep their arguments (size) in a structure (like a call-stack) so the lengths can be properly resolved.
May be we would need to keep this in the SSA pass?

Reply:
This function is not called when the frontend is parsing, it is called during monomorphisation. Good catch though, this was a bug rather than something inherently wrong with the pass. I've went ahead and fixed the bug and added the extra test function which clearly should have been there from the start.

The bug itself was the monomorphisation pass needed to call `Type::follow_bindings` on each calls type bindings so that they would not be altered when other functions in the function queue push and pop their bindings. There was also a bug within follow_bindings itself where NamedGenerics weren't being followed like other type variables were.